state/runtime/evm/state.go

```diff
@@ -210,30 +213,84 @@ func (c *state) resetReturnData() {
 
 // Run executes the virtual machine
 func (c *state) Run() ([]byte, error) {
-	var vmerr error
+	var (
+		vmerr  error
+		logged bool
+
+		op      OpCode
+		ipCopy  int
+		gasCopy uint64
+
+		ok bool
+	)
+
+	tracer := c.host.GetTracer()
+
+	defer func() {
+		if vmerr == nil || tracer == nil {
+			return
+		}
+
+		if !logged {
+			tracer.ExecuteState(
+				c.msg.Address,
+				ipCopy,
+				op.String(),
+				gasCopy,
+				// XXX: c.currentConsumedGas is not the total of gas assumption in a instruction
+				// when the execution terminated in the middle
+				c.currentConsumedGas,
+				c.returnData,
+				c.msg.Depth,
+				c.err,
+				c.host,
+			)
+		} else {
+			tracer.ExecuteFault(
```

Comment:
Can we merge `ExecuteFault` and `ExecuteState`?

Reply:
I removed `ExecutionFault` for now because struct tracer doesn't need it